trieste/acquisition/function/greedy_batch.py

```diff
@@ -351,3 +355,295 @@ def __call__(self, x: TensorType) -> TensorType:
+        self,
+        models: Mapping[str, ProbabilisticModel],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+
+        :param models: The models over each tag.
+        :param datasets: The data from the observer (optional).
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+        if pending_points is not None:
+            tf.debugging.assert_rank(pending_points, 2)
+
+            fantasized_data = {
+                tag: _generate_fantasized_data(
+                    fantasize_method=self._fantasize_method,
+                    model=model,
+                    pending_points=pending_points,
+                )
+                for tag, model in models.items()
+            }
+            models = {
+                tag: _fantasize_model(model, fantasized_data[tag]) for tag, model in models.items()
+            }
+
+            if datasets is None:
+                datasets = fantasized_data
+            else:
+                datasets = {tag: data + fantasized_data[tag] for tag, data in datasets.items()}
+        return self._builder.prepare_acquisition_function(models, datasets)
+
+
+def _generate_fantasized_data(
+    fantasize_method: str, model: ProbabilisticModel, pending_points: TensorType
+) -> Dataset:
+    """
+    Generates "fantasized" data at pending_points depending on the chosen heuristic:
+    - KB (kriging believer) uses the mean prediction of the models
+    - sample uses samples from the GP posterior.
+
+    :param fantasize_method: one of "KB", "sample"
+    :param model: a model with predict method
+    :param dataset: past data
+    :param pending_points: points at which to fantasize data
+    :return: a fantasized dataset
+    """
+    if fantasize_method == "KB":
+        fantasized_obs, _ = model.predict(pending_points)
+    elif fantasize_method == "sample":
+        fantasized_obs = model.sample(pending_points, num_samples=1)[0]
+    else:
+        raise NotImplementedError(
+            f"fantasize_method must be KB or sample, " f"received {model.__repr__()}"
+        )
+
+    return Dataset(pending_points, fantasized_obs)
+
+
+def _fantasize_model(
+    model: ProbabilisticModel, fantasized_data: Dataset
+) -> _fantasized_model | ModelStack:
+    if isinstance(model, ModelStack):
+        observations = tf.split(fantasized_data.observations, model._event_sizes, axis=-1)
+        fmods = []
+        for mod, obs, event_size in zip(model._models, observations, model._event_sizes):
+            fmods.append(
+                (
+                    _fantasized_model(mod, Dataset(fantasized_data.query_points, obs)),
+                    event_size,
+                )
+            )
+        return ModelStack(*fmods)
```

Comment:
this line generates the following message at tox -e types:

error: Argument 1 to "ModelStack" has incompatible type "*List[Tuple[_fantasized_model, Any]]"; expected "Tuple[TrainableProbabilisticModel, int]"

I have no idea how to fix this neatly. @uri-granta ??

Reply:
I think it's complaining that `_fantasized_model` isn't a TrainableProbabilisticModel just a ProbabilisticModel, and ModelStack currently expects the former (so it can call update and optimize).

Not sure what the nicest solution is. Perhaps to split ModelStack into a `NonTrainableModelStack(ProbabilisticModel)` class that supports all models and a `ModelStack(NonTrainableModelStack, TrainableProbabilisticModel)` subclass that only supports trainable models (like the current version) and adds the update and optimize methods?

Reply:
Possibly `ModelStack` and `TrainableModelStack`?

The alternative would be to have dummy trainable methods for `fantasized_models` - a bit hacky but less code change?

I don't have a strong opinion either way...

Reply:
went for splitting ModelStack